src/levels.rs

```diff
@@ -1,5 +1,1053 @@
+        compact_def.next_range = KeyRange::default();
+        compact_def.bot = vec![];
+
+        let this_level = compact_def.this_level.read().unwrap();
+        // Since next_level and this_level are same L0, do not need to acquire lock of next level.
+        // TODO: Don't lock cpt_status through this function.
+        let mut cpt_status = self.cpt_status.write().unwrap();
+
+        let mut out = vec![];
+
+        for table in this_level.tables.iter() {
+            if table.size() >= 2 * compact_def.targets.file_size[0] {
+                // file already big, don't include it
+                continue;
+            }
+
+            // TODO: Add created at logic.
+
+            if cpt_status.tables.contains(&table.id()) {
+                continue;
+            }
+            out.push(table.clone());
+        }
+
+        if out.len() < 4 {
+            return Err(Error::CustomError("not enough table to merge".to_string()));
+        }
+
+        compact_def.this_range = KeyRange::Inf;
+        compact_def.top = out;
+
+        // Avoid any other L0 -> Lbase from happening, while this is going on.
+        // TODO: ?
+        cpt_status.levels[this_level.level]
+            .ranges
+            .push(KeyRange::Inf);
+        for table in compact_def.top.iter() {
+            assert!(!cpt_status.tables.insert(table.id()));
+        }
+
+        // Make the output always one file to decreases the L0 table stalls and
+        // improves the performance.
+        compact_def.targets.file_size[0] = std::u32::MAX as u64;
+
+        Ok(())
+    }
+
+    fn fill_tables_l0_to_lbase(&self, compact_def: &mut CompactDef) -> Result<()> {
+        if compact_def.next_level_id == 0 {
+            panic!("base level can't be zero");
+        }
+
+        let this_level = compact_def.this_level.read().unwrap();
+        let next_level = compact_def.next_level.read().unwrap();
+
+        if compact_def.prios.adjusted > 0.0 && compact_def.prios.adjusted < 1.0 {
+            return Err(Error::CustomError(
+                "score less than 1.0, not compact to Lbase".to_string(),
+            ));
+        }
+
+        if this_level.tables.is_empty() {
+            return Err(Error::CustomError("no table in this level".to_string()));
+        }
+
+        let mut out = vec![];
+
+        if !compact_def.drop_prefixes.is_empty() {
+            out = this_level.tables.clone();
+        } else {
+            let mut kr = KeyRange::default();
+            // Start from the oldest file first.
+            for table in this_level.tables.iter() {
+                let dkr = get_key_range_single(table);
+                if kr.overlaps_with(&dkr) {
+                    out.push(table.clone());
+                    kr = kr.extend(&dkr);
+                } else {
+                    break;
+                }
+            }
+        }
+        compact_def.this_range = get_key_range(&compact_def.top);
+        compact_def.top = out;
```

Comment:
Why not place L670 before L669 since L672 using the `compact_def.this_range` to calculate the overlap files?

Reply:
That's a mistake, I will fix it, thanks.

Reply:
Seems like because of `Empty` range overlapping with any range, we could pass all tests. I will try to add test for this.